darts/models/forecasting/forecasting_model.py

```diff
@@ -431,15 +461,27 @@ def historical_forecasts(
         last_points_values = []
 
         # iterate and forecast
-        for pred_time in iterator:
+        for _counter, pred_time in enumerate(iterator):
             # build the training series
             train = series.drop_after(pred_time)
             if train_length and len(train) > train_length:
                 train = train[-train_length:]
 
             # train_cov = covariates.drop_after(pred_time) if covariates else None
 
-            if retrain or not self._fit_called:
+            if (not self._fit_called) or retrain_func(
+                counter=_counter,
+                pred_time=pred_time,
+                series=train,
+                past_covariates=past_covariates.drop_after(pred_time)
+                if past_covariates
+                else None,
+                future_covariates=future_covariates.drop_after(
```

Comment:
This will fail in some cases (and causes the quickstart notebook to fail), because the actual forecasts do not go further than the end of `series` in backtesting. So you should probably do something like
```python
future_covariates=future_covariates.drop_after(
    min(pred_time + series.freq * forecast_horizon, series.end_time()))
```